lib/view_component/compiler.rb

```diff
@@ -38,46 +34,39 @@ def compile(raise_errors: false, force: false)
       end
 
       component_class.superclass.compile(raise_errors: raise_errors) if should_compile_superclass?
+      subclass_instance_methods = component_class.instance_methods(false)
 
-      with_write_lock do
-        CompileCache.invalidate_class!(component_class)
-
-        subclass_instance_methods = component_class.instance_methods(false)
-
-        if subclass_instance_methods.include?(:with_content) && raise_errors
-          raise ViewComponent::ComponentError.new(
-            "#{component_class} implements a reserved method, `#with_content`.\n\n" \
-            "To fix this issue, change the name of the method."
-          )
-        end
-
-        if template_errors.present?
-          raise ViewComponent::TemplateError.new(template_errors) if raise_errors
+      if subclass_instance_methods.include?(:with_content) && raise_errors
+        raise ViewComponent::ComponentError.new(
+          "#{component_class} implements a reserved method, `#with_content`.\n\n" \
+          "To fix this issue, change the name of the method."
+        )
+      end
 
-          return false
-        end
+      if template_errors.present?
+        raise ViewComponent::TemplateError.new(template_errors) if raise_errors
 
-        if subclass_instance_methods.include?(:before_render_check)
-          ViewComponent::Deprecation.warn(
-            "`#before_render_check` will be removed in v3.0.0.\n\n" \
-            "To fix this issue, use `#before_render` instead."
-          )
-        end
+        return false
+      end
 
-        if raise_errors
-          component_class.validate_initialization_parameters!
-          component_class.validate_collection_parameter!
-        end
+      if subclass_instance_methods.include?(:before_render_check)
+        ViewComponent::Deprecation.warn(
+          "`#before_render_check` will be removed in v3.0.0.\n\n" \
+          "To fix this issue, use `#before_render` instead."
+        )
+      end
 
-        templates.each do |template|
-          # Remove existing compiled template methods,
-          # as Ruby warns when redefining a method.
-          method_name = call_method_name(template[:variant])
+      if raise_errors
+        component_class.validate_initialization_parameters!
+        component_class.validate_collection_parameter!
+      end
 
-          if component_class.instance_methods.include?(method_name.to_sym)
-            component_class.send(:undef_method, method_name.to_sym)
-          end
+      templates.each do |template|
+        # Remove existing compiled template methods,
+        # as Ruby warns when redefining a method.
+        method_name = call_method_name(template[:variant])
 
+        silence_warnings do
```

Comment:
```suggestion
        method_name = call_method_name(template[:variant])
        component_class.silence_redefinition_of_method(method_name)
```

Thoughts on using the [ActiveSupport `silence_redefinition_of_method`](https://www.rubydoc.info/gems/activesupport/Module:silence_redefinition_of_method) method instead here? That would let us keep any non-redefinition warnings here, and maybe be a bit more intention revealing.

Reply:
I'm all for being more specific!

Reply:
Oh nice!

Reply:
Unfortunately tests failed using the `silence_redefinition_of_method` approach because it's not threadsafe 🤦 It first checks to see if the method in question is defined, then cleverly uses a side-effect of `alias_method` to define the method. Unfortunately the method might not exist during the check but exist immediate after it, meaning the `class_eval` will still result in a warning. I pushed up a threadsafe version that passes all our tests, including the ones that use threads.

Reply:
I think the solution you has will work, but it's tracking per-thread which seems unnecessary?

It feels like a scenario where we'd want to actually use a lock, which I think would be safe since it could be tracked as internal compiler state?

I'll add a comment in the code too real quick.